Add more suggestions for avoiding deadlocks to webhook docs (#46798)
kubernetes/website: content/en/docs/reference/access-authn-authz/extensible-admission-controllers.md

### Avoiding deadlocks in self-hosted webhooks

There are several ways that webhooks can cause deadlocks, where the cluster cannot make progress in
scheduling pods:

* A webhook running inside the cluster might cause deadlocks for its own deployment if it is configured
  to intercept resources required to start its own pods.

  For example, a mutating admission webhook is configured to admit `CREATE` pod requests only if a certain label is set in the
  pod (e.g. `"env": "prod"`). The webhook server runs in a deployment which doesn't set the `"env"` label.
  When a node that runs the webhook server pods
  becomes unhealthy, the webhook deployment will try to reschedule the pods to another node. However the requests will
  get rejected by the existing webhook server since the `"env"` label is unset, and the migration cannot happen.

  It is recommended to exclude the namespace where your webhook is running with a

>   [namespaceSelector](#matching-requests-namespaceselector).

Reviewer: Let's mention ValidatingAdmissionPolicies as a way to provide (at least) some protection without having the admission webhook protect itself.

Author: Added a few lines down.

* If the cluster has multiple webhooks configured (possibly from independent applications deployed on
  the cluster), they can form a cycle.  Webhook A must be called to process startup of webhook B's
  pods and vice versa.  If both webhook A and webhook B ever become inactive at the same time (for
  example, due to a cluster-wide outage or a node failure where both pods run on the same node)
  deadlock occurs because neither webhook pod can be recreated without the other already running.

  It is recommended to establish a desired order for webhooks to start, then to exclude "earlier"
  webhooks' resources from being inspected by later webhooks.  This ensures that the "earliest"
  webhook can start, which in turn allows the next.

* Webhooks can intercept rsesources used by critical cluster infrastructure, such as coredns,
  the active CNI and, storage plugins.  These resources may be required to schedule the webhook pods
  on the cluster, creating a deadlock if the webhook is configured to intercept them.

  It is recommended to exclude cluster infratructure namespaces from webhooks, including kube-system,

>   any namespaces used by CNI plugins, etc.

Reviewer: - typo (“infratructure”)
- let's mention ValidatingAdmissionPolicies as a way to provide (at least) some protection without having the admission webhook protect itself

### Avoiding deadlocks in self-hosted webhooks

There are several ways that webhooks can cause deadlocks, where the cluster cannot make progress in
scheduling pods:

* A webhook running inside the cluster might cause deadlocks for its own deployment if it is configured
  to intercept resources required to start its own pods.

>   For example, a mutating admission webhook is configured to admit `CREATE` pod requests only if a certain label is set in the

Reviewer: ```suggestion
  For example, a mutating admission webhook is configured to admit **create** Pod requests only if a certain label is set in the
```

>   pod (such as `"env": "prod"`). The webhook server runs in a deployment which doesn't set the `"env"` label.

Reviewer: ```suggestion
  pod (such as `env: "prod"`). However, the webhook server runs as a Deployment that doesn't set the `env` label.
```

  When a node that runs the webhook server pods
  becomes unhealthy, the webhook deployment will try to reschedule the pods to another node. However the requests will

>   get rejected by the existing webhook server since the `"env"` label is unset, and the migration cannot happen.

Reviewer: ```suggestion
  get rejected by the existing webhook server since the `env` label is unset, and the replacement Pod
  cannot be created. Eventually, the entire Deployment for the webhook server may become unhealthy.
```

>   It is recommended to exclude the namespace where your webhook is running with a

Reviewer: This isn't a K8s recommendation. Using a `namespaceSelector` is one good option.
```suggestion
If you use admission webhooks to check Pods, consider excluding the namespace where your webhook
listener is running, by specifying a
```

  [namespaceSelector](#matching-requests-namespaceselector).

* If the cluster has multiple webhooks configured (possibly from independent applications deployed on
  the cluster), they can form a cycle.  Webhook A must be called to process startup of webhook B's
  pods and vice versa.  If both webhook A and webhook B ever become inactive at the same time (for
  example, due to a cluster-wide outage or a node failure where both pods run on the same node)
  deadlock occurs because neither webhook pod can be recreated without the other already running.

>   It is recommended to establish a desired order for webhooks to start, then to exclude "earlier"

Reviewer: I don't think this is a recommendation that the Kubernetes project makes.

Author: OK, I've toned it down from recommendation to an example

  webhooks' resources from being inspected by later webhooks.  This ensures that the "earliest"
  webhook can start, which in turn allows the next.

>   Where protection for the webhook and/or its namespace is desired, [Validating Admission Policies](/docs/reference/access-authn-authz/validating-admission-policy/)

Reviewer: ```suggestion
  If you want to ensure protection for an admission webhook and / or its namespace, [ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
  can
```

  provide many protection capabilities without introducing dependency cycles.

> * Webhooks can intercept rsesources used by critical cluster infrastructure, such as coredns,

Reviewer: ```suggestion
* Admission webhooks can intercept resources used by critical cluster add-ons, such as CoreDNS,
```

>   the active CNI and, storage plugins.  These resources may be required to schedule the webhook pods

Reviewer: ```suggestion
  network plugins, or storage plugins. These add-ons may be required to schedule or successfully run the
  pods for a particular admission webhook
```

  on the cluster, creating a deadlock if the webhook is configured to intercept them.

  It is recommended to exclude cluster infrastructure namespaces from webhooks, including kube-system,

>   any namespaces used by CNI plugins, etc.

Reviewer: This isn't a Kubernetes project recommendation, although in many (not all) cases it is a good idea to do.

Author: toned down.

>   Where protection for the infrastructure components is desired, [Validating Admission Policies](/docs/reference/access-authn-authz/validating-admission-policy/)

Reviewer: ```suggestion
  If you want to ensure protection for an core add-on / or its namespace,
  [ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
  can
```

>   provide many protection capabilities without introducing dependency cycles.

Reviewer: ```suggestion
  provide many protection capabilities without any dependency on worker nodes and Pods.
```

> ### Avoiding deadlocks in self-hosted webhooks

Reviewer 2: Maybe this scenario can also be incorporated: https://github.com/kubernetes/enhancements/pull/5068#discussion_r1945419681

Author: This PR feels like it's at danger of not converging, I'd rather not add anything more to it.

  If you use admission webhooks to check Pods, consider excluding the namespace where your webhook
  listener is running, by specifying a
  [namespaceSelector](#matching-requests-namespaceselector).

* If the cluster has multiple webhooks configured (possibly from independent applications deployed on
  the cluster), they can form a cycle.  Webhook A must be called to process startup of webhook B's
  pods and vice versa. If both webhook A and webhook B ever become unavailable at the same time (for
  example, due to a cluster-wide outage or a node failure where both pods run on the same node)
  deadlock occurs because neither webhook pod can be recreated without the other already running.

  One way to prevent this is to exclude webhook A's pods from being acted on be webhook B. This
  allows webhook A's pods to start, which in turn allows webhook B's pods to start. If you had a 
  third webhook, webhook C, you'd need to exclude both webhook A and webhook B's pods from 
  webhook C. This ensures that webhook A can _always_ start, which then allows webhook B's pods 
  to start, which in turn allows webhook C's pods to start.

>   If you want to ensure protection for an admission webhook and / or its namespace, [ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
>   can
>   provide many protection capabilities without introducing dependency cycles.

Reviewer: Oops, this is technically incorrect.
```suggestion
  If you want to ensure protection that avoids these risks, [ValidatingAdmissionPolicies](/docs/reference/access-authn-authz/validating-admission-policy/)
  can
  provide many protection capabilities without introducing dependency cycles.
```

  get rejected by the existing webhook server since the `env` label is unset, and the replacement Pod
  cannot be created. Eventually, the entire Deployment for the webhook server may become unhealthy.

  If you use admission webhooks to check Pods, consider excluding the namespace where your webhook
  listener is running, by specifying a
  [namespaceSelector](#matching-requests-namespaceselector).

* If the cluster has multiple webhooks configured (possibly from independent applications deployed on
  the cluster), they can form a cycle.  Webhook A must be called to process startup of webhook B's
  pods and vice versa. If both webhook A and webhook B ever become unavailable at the same time (for
  example, due to a cluster-wide outage or a node failure where both pods run on the same node)
  deadlock occurs because neither webhook pod can be recreated without the other already running.

  One way to prevent this is to establish a desired order for webhooks to start, then to exclude 
  "earlier" webhooks' resources from being inspected by "later" webhooks.  This ensures that the 

>   "earliest" webhook can start, which in turn allows the next.

Reviewer: I don't understand how this could work and don't think it's an effective control. See https://stackoverflow.com/a/69166466 for informal confirmation that admission webhooks are _not_ ordered.

Author: I think you've taken the wrong idea from that paragraph, which suggests I could explain it more clearly!  I'm talking about the order in which the webhook pods start rather than the order that webhooks apply to a particular resource.

The deadlock that can happen is if 

* webhook A is unavailable
* webhook B is unavailable
* webhook A is needed to approve webhook B's pod
* webhook B is needed to approve webhook A's pod.

So the suggestion is:

* Pick the webhook that needs to apply most broadly.  Let's say that's A
* Make sure that no other webhooks can interfere with starting A.
* Therefore A can always start, even if no other webhooks are available
* Therefore B can start, because it depends on A but we know that A can start.

If you have3 webhooks, A, B, C, then

* Make sure B and C exclude A's pods
* Therefore A can start
* Make sure C excludes B's pods
* Therefore B can start, since we know that A can start
* Therefore C can start, since A and B can start.

Author: I'll try to reword it, but suggestions also welcome

Reviewer: This kind of workload sequencing doesn't really hold because nodes are not considered durable and resilient. Also, do try to reword.

Reviewer: I get the point we want to make. Maybe that's worth shifting to its own point (eg a blog article)?

Author: I think as long as you've got an ordering, node failure is fine:
* webhook A comes up
* webhook A's node dies
* webhook A is rescheduled somewhere else
* now webhook B can start... and so on.

Author: It should get there _eventually_, whereas without some sort of ordering like this, node failures will cause deadlocks and you'll be forced to intervene by temporarily removing your webhook
